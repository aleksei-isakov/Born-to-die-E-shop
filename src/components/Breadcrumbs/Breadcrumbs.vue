<template>
  <div v-if="isShown" class="breadcrumbs__wrapper page-container">
    <v-breadcrumbs :items="breadCrumbs">
      <template v-slot:item="{ item }">
        <v-breadcrumbs-item :to="item.to" exact>
          <span class="breadcrumb__text">{{ item.text }}</span>
        </v-breadcrumbs-item>
      </template>
    </v-breadcrumbs>
  </div>
</template>

<script>
import { mapGetters, mapActions } from 'vuex';

export default {
  name: 'Breadcrumbs',

  computed: {
    ...mapGetters('PdpPageModule', ['productInfo']),

    breadCrumbs() {
      if (typeof this.$route.meta.breadCrumb === 'function') {
        let result = this.$route.meta.breadCrumb.call(this, this.$route);

        return result.map((res) => {
          if (res.params) {
            res.text = this.productInfo.name;

            return res;
          }

          return res;
        });
      }

      return this.$route.meta.breadCrumb;
    },

    isShown() {
      return this.$route.name !== 'Home';
    }
  },

  async mounted() {
    if (this.$route.params.id) {
      await this.getProductInfo(this.$route.params.id);
    }
  },

  methods: {
    ...mapActions('PdpPageModule', ['getProductInfo'])
  }
};
</script>

<style lang="scss" scoped>
@import '@/scss/CustomVariables.scss';

.breadcrumbs__wrapper {
  ul {
    padding-left: 0;
  }
}

.breadcrumbs__wrapper {
  text-decoration: none;
}

.breadcrumbs__wrapper::v-deep a:hover {
  text-decoration: none;
}

.breadcrumb__text {
  color: $font-color-text;
}
</style>
