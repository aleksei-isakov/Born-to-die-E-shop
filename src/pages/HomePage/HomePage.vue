<template>
  <div>
    <PLPSearchBar />
    <ProductsList :products="productsList" />
  </div>
</template>

<script>
import ProductsList from '@/components/ProductsList/ProductsList.vue';
import PLPSearchBar from '@/components/PLPSearchBar/PLPSearchBar.vue';
import { mapGetters, mapActions } from 'vuex';

export default {
  name: 'HomePage',

  components: {
    ProductsList,
    PLPSearchBar
  },

  computed: {
    ...mapGetters('ProductsModule', ['productsList'])
  },

  async mounted() {
    await this.getProductsList();
  },

  methods: {
    ...mapActions('ProductsModule', ['getProductsList'])
  }
};
</script>
