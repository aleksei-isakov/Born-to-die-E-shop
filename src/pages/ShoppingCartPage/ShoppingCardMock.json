[
  {
    "id": "45196514-a1d2-45ba-8e8d-c49e60cc99c3",
    "name": "Unbranded Steel Bike",
    "category": {
      "id": "e42374a2-7f13-4a5e-8723-20b02908bab3",
      "name": "nobis",
      "description": "Corporis sunt voluptatibus veniam corporis esse ea dicta.",
      "createdAt": "2021-06-01T05:48:34.840Z",
      "updatedAt": ""
    },
    "description": "The Football Is Good For Training And Recreational Purposes",
    "price": 7,
    "images": [
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480"
    ],
    "createdAt": "2021-03-27T18:39:19.237Z",
    "updatedAt": ""
  },
  {
    "id": "8a45ebaf-a3c7-468b-8cfb-d0b4147d01d6",
    "name": "Incredible Frozen Keyboard",
    "category": {
      "id": "0f515427-5d29-4e8b-bead-09bf17275649",
      "name": "sed",
      "description": "Rerum eum nobis facilis impedit quos non.",
      "createdAt": "2021-06-27T00:26:52.257Z",
      "updatedAt": ""
    },
    "description": "The slim & simple Maple Gaming Keyboard from Dev Byte comes with a sleek body and 7- Color RGB LED Back-lighting for smart functionality",
    "price": 679,
    "images": [
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480"
    ],
    "createdAt": "2021-06-30T06:53:12.692Z",
    "updatedAt": ""
  },
  {
    "id": "9d7732e2-116f-4557-a32a-53200786f8d1",
    "name": "Unbranded Fresh Mouse",
    "category": {
      "id": "fd82d3d4-1475-4da6-b8e9-28663cd8df8a",
      "name": "dicta",
      "description": "Temporibus est quisquam aut vitae numquam similique.",
      "createdAt": "2022-01-07T15:09:54.594Z",
      "updatedAt": ""
    },
    "description": "The slim & simple Maple Gaming Keyboard from Dev Byte comes with a sleek body and 7- Color RGB LED Back-lighting for smart functionality",
    "price": 579,
    "images": [
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480"
    ],
    "createdAt": "2021-09-09T23:17:40.690Z",
    "updatedAt": ""
  },
  {
    "id": "431059dd-e258-4962-b528-3e5156996fd7",
    "name": "Generic Plastic Gloves",
    "category": {
      "id": "fd82d3d4-1475-4da6-b8e9-28663cd8df8a",
      "name": "dicta",
      "description": "Temporibus est quisquam aut vitae numquam similique.",
      "createdAt": "2022-01-07T15:09:54.594Z",
      "updatedAt": ""
    },
    "description": "The automobile layout consists of a front-engine design, with transaxle-type transmissions mounted at the rear of the engine and four wheel drive",
    "price": 48,
    "images": [
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480",
      "http://placeimg.com/640/480"
    ],
    "createdAt": "2022-01-28T19:50:03.615Z",
    "updatedAt": ""
  }
]

