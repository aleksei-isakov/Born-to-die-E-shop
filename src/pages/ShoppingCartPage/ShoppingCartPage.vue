<template>
  <div class="shopping-cart-page__wrapper page-container">
    <div>
      <empty-cart-popup
        v-if="isPopupVisible"
        @close="onClickTogglePopup"
        @clear-cart="onClickClearShoppingCart"
      />
    </div>

    <h1 class="shopping-cart-page__title">Cart</h1>

    <div v-if="isCartEmpty" class="cart-empty-container">
      <p class="cart-empty-container__message">
        No one product has been added to the card yet : (
        <br />
        <base-button-router path="/products">
          Start searching
        </base-button-router>
      </p>
    </div>

    <div v-else>
      <empty-cart-button
        class="shopping-cart-page__empty-cart-button"
        :disabled="isDisabled"
        @show-popup="onClickTogglePopup"
      />

      <shopping-cart-list
        class="shopping-cart-page__product-list-container"
        :products-in-cart="productsInCart"
        :total-price="totalPrice"
      />
    </div>
    <shopping-cart-list
      :products-in-cart="productsInCart"
      :total-price="totalPrice"
      :checked-products="checkedProductsInCart"
    />
  </div>
</template>

<script>
import ShoppingCartList from '@/components/ShoppingCartList/ShoppingCartList';
import EmptyCartButton from '@/components/EmptyCartButton/EmptyCartButton';
import EmptyCartPopup from '@/components/EmptyCartPopup/EmptyCartPopup';
import { mapGetters, mapActions } from 'vuex';
import { BaseButtonRouter } from '@/base_components';

export default {
  name: 'ShoppingCartPage',

  components: {
    BaseButtonRouter,
    ShoppingCartList,
    EmptyCartButton,
    EmptyCartPopup
  },

  data() {
    return {
      isPopupVisible: false,
      isDisabled: false
    };
  },

  computed: {
    ...mapGetters('ShoppingCartModule', [
      'productsInCart',
      'totalPrice',
      'checkedProductsInCart'
    ]),

    isCartEmpty() {
      return this.productsInCart.length === 0;
    }
  },

  methods: {
    ...mapActions('ShoppingCartModule', ['clearCart']),

    onClickTogglePopup() {
      this.isPopupVisible = !this.isPopupVisible;
    },

    onClickClearShoppingCart() {
      this.clearCart();
      this.isPopupVisible = false;
      this.isDisabled = true;
    }
  }
};
</script>

<style lang="scss" scoped>
@import '@/scss/CustomVariables.scss';

.shopping-cart-page {
  &__wrapper {
    position: relative;
  }

  &__empty-cart-button {
    position: absolute;
    right: 20px;
    top: 20px;
  }

  &__title {
    color: $font-color-subtitle;
    font-weight: normal;
    align-self: flex-start;
    margin-bottom: 25px;
  }
}

.cart-empty-container {
  display: flex;
  flex-direction: column;

  &__message {
    font-size: $font-size-subtitle;
    color: $font-color-subtitle;

    .base-button-router {
      color: $primary;

      &:hover {
        color: $primary;
      }
    }
  }
}
</style>
