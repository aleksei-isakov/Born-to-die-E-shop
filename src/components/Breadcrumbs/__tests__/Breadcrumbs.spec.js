import { mount, createLocalVue } from '@vue/test-utils';
import Vue from 'vue';
import VueRouter from 'vue-router';
import Vuetify from 'vuetify';
import Breadcrumbs from '@/components/Breadcrumbs/Breadcrumbs';
import ShoppingCartPage from '@/pages/ShoppingCartPage/ShoppingCartPage';
import HomePage from '@/pages/HomePage/HomePage';
Vue.use(Vuetify);

const routes = [
  {
    path: '/',
    name: 'Home',
    component: HomePage,
    meta: {
      breadCrumb: [{ text: 'Home' }]
    }
  },
  {
    path: '/cart',
    name: 'ShoppingCartPage',
    component: ShoppingCartPage,
    meta: {
      breadCrumb: [{ text: 'Home', to: { name: 'Home' } }, { text: 'Cart' }]
    }
  }
];
const localVue = createLocalVue();

localVue.use(VueRouter);
const router = new VueRouter({ routes });
let wrapper;

beforeEach(() => {
  wrapper = mount(Breadcrumbs, {
    localVue,
    router
  });

  router.push('/cart').catch((error) => {
    if (
      error.name !== 'NavigationDuplicated' &&
      !error.message.includes(
        'Avoided redundant navigation to current location'
      )
    ) {
      console.log(error);
    }
  });
});

afterEach(() => {
  wrapper.destroy();
});

describe('Breadcrumbs', () => {
  it('renders a valid snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('return to home page', () => {
    wrapper.find('a').trigger('click');
    expect(wrapper.vm.$route.path).toBe('/');
  });

  it('should contain breadcrumbs container', () => {
    expect(wrapper.find('.breadcrumbs__wrapper').exists()).toBe(true);
  });
});
