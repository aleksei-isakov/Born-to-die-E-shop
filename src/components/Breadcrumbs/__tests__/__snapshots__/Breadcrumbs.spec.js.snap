// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumbs renders a valid snapshot 1`] = `
<div class="breadcrumbs__wrapper container">
  <ul class="v-breadcrumbs theme--light">
    <li><a href="#/" class="v-breadcrumbs__item"><span class="breadcrumb__text">Home</span></a></li>
    <li class="v-breadcrumbs__divider">/</li>
    <li>
      <div class="v-breadcrumbs__item"><span class="breadcrumb__text">Cart</span></div>
    </li>
  </ul>
</div>
`;
